Added missing docs for prepareParams to Relay.Route.

Globegitter opened the pull request:
Given the issue https://github.com/facebook/relay/issues/1101 and that there are no docs for `prepareParams` yet I thought I could do both at once.

I haven't actually tested this example out as such, but assuming that `prepareParams` in `react-router-relay` (https://github.com/relay-tools/react-router-relay/tree/6c36030e8bdfba7ca2e344a6436135610b6e2256#additional-parameters) do behave exactly the same as stated.

Globegitter commented:
I know this example is a bit overloaded, since it actually covers bits of the `Route` as well as the `Relay.Container` but it does kind of fit and would have at helped us when we were trying to solve that specific use-case.

josephsavona commented on docs/APIReference-Route.md:
```diff
@@ -70,6 +76,72 @@ class ProfileRoute extends Relay.Route {
+    limit: null
+  },
+  fragments: {
+    viewer: ({limit}) => Relay.QL`
```
```
viewer: () => Relay.QL`...`
```

The `limit` argument is unnecessary here.

josephsavona replied:
ping

josephsavona commented on docs/APIReference-Route.md:
```diff
@@ -70,6 +76,72 @@ class ProfileRoute extends Relay.Route {
+    viewer: () => Relay.QL`query { viewer }`
+  };
+  static prepareParams = (prevParams) => {
+    // Make sure any params that ProfileRoute gets initialised with will still be passed down.
```
nit: move this comment before or to the end of the `...prevParams` line? it's ambiguous what it's referring to here.

josephsavona commented on docs/APIReference-Route.md:
```diff
@@ -70,6 +76,72 @@ class ProfileRoute extends Relay.Route {
+    // Make sure any params that ProfileRoute gets initialised with will still be passed down.
+    return {
+      ...prevParams,
+      limit: 10
```
to help clarify the other comment, it might be good to comment this line too, maybe something like "and override the `limit` variable in the top level container.

josephsavona commented on docs/APIReference-Route.md:
````diff
@@ -70,6 +76,72 @@ class ProfileRoute extends Relay.Route {
+});
+```
+In this example query parameters are being passed down to the `Child` component: One comes from the `ProfileRoute` via `prepareParams` and given `ProfileRoute` and `Container` are both being passed into `Relay.Renderer`, it automatically passed it down. The `Container` then passes two different `limit` variables down to the `Child` one from the `Route` and one from within.
+
````
It's worth documenting this, but it doesn't really have to do with routes or prepareParams. What about adding a section to the RelayContainer guide (linked to from the Route guide) about variable passing?

Globegitter replied:
Yeah that is a good idea, will update it that way asap.

Globegitter commented:
@josephsavona Have addressed most of your comments. Seeing as you have now more thoroughly documented the variable passing down (https://github.com/facebook/relay/pull/1142) I simplified the example.

I did however think that some of the other sections could do with some further explanation. For me it is often more easy to understand things based on an (explained) example.

Let me know if anything else needs fixing up.

akre54 commented on docs/APIReference-Route.md:
```diff
@@ -70,6 +76,57 @@ class ProfileRoute extends Relay.Route {
+
+const Container = Relay.createContainer(({viewer}) => (
+  <View>
+    {viewer.edges.map((node) => <div>node.name</div>)}
```
Shouldn't this be `viewer.edges.map((edge) => <div>{edge.node.name}</div>)`?

Globegitter replied:
Thanks for the catch. The one disadvantage of being fullstack and not having been able to write any meaningful JS in about a month.

Globegitter commented:
@josephsavona Anything still blocking this PR being merged?

josephsavona commented:
@Globegitter see my comment about the extraneous `limit` argument

josephsavona commented:
otherwise looks good

Globegitter commented:
Somehow missed. Just updated.

josephsavona commented on docs/APIReference-Route.md:
````diff
@@ -70,6 +76,57 @@ class ProfileRoute extends Relay.Route {
+});
+```
+In this example `prepareParams` overrides any `limit` variables the route gets initialised with. Given `ProfileRoute` and `Container` are both being passed into `Relay.Renderer`, that limit variable will automatically be passed down to `Container`.
+
````
@Globegitter on second reading, this example conflates two things: `prepareParams` and passing route vars to the root container fragment.

Can you separate these? Note that `prepareParams` can be used for things like setting default values or converting from strings (e.g. in a URL) to integers. I'd recommend keeping this PR focused on `prepareParams` only, and maybe add a follow-up for the route -> root container variable example.

wincent commented:
Going to pick this up and polish off any pending details necessary to get this through internal code-review. Thanks, @Globegitter.

wincent commented:
@facebook-github-bot import

facebook-github-bot-3 commented:
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/503983499788689/int_phab) to review.

wincent commented:
@Globegitter: I landed the modified version of this, incorporating @josephsavona's suggestions:

> Can you separate these? Note that `prepareParams` can be used for things like setting default values or converting from strings (e.g. in a URL) to integers. I'd recommend keeping this PR focused on `prepareParams` only, and maybe add a follow-up for the route -> root container variable example.

Globegitter commented:
@wincent Sorry for the delay and thanks for making the last changes.
